### PR TITLE
EdgeAgent: Fail fast if createOptions are not valid

### PR DESCRIPTION
**Pull request by varunpuranik**
- Added check to fail fast if createOptions in deployment are not valid
- Also added check to add "latest" tag to image name if no tag is specified

### CODE REVIEW

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/Constants.cs [hunk 1]
```diff
@@ -3,7 +3,9 @@ namespace Microsoft.Azure.Devices.Edge.Agent.Docker
 {
     static class Constants
     {
-        public const string EdgeAgentCreateOptionsName = "EdgeAgentCreateOptions";
+        public const string UnknownImage = "unknown";
```
Comment (mrohera) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/Constants.cs]: LGTM overall. Suggest changing the value to "unknown-module-docker-image" or something along those lines to make the logs more assertive and actionable.

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs [hunk 2]
```diff
@@ -25,6 +28,10 @@ public class DockerConfigTest
 
         static readonly DockerConfig Config10 = new DockerConfig("image1:42", @"{""Env"": [""k11=v11"", ""k22=v22""], ""HostConfig"": {""PortBindings"": {""43/udp"": [{""HostPort"": ""43""}]}}}");
         static readonly DockerConfig Config11 = new DockerConfig("image1:42", @"{""Env"": [""k33=v33"", ""k44=v44""], ""HostConfig"": {""PortBindings"": {""43/udp"": [{""HostPort"": ""43""}]}}}");
+        static readonly DockerConfig Config12 = new DockerConfig("image1:42", string.Empty);
```
Comment (mrohera) [inline edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs]: Also suggest adding a test for unknown image for completeness.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/Constants.cs (hunk 1 above)
Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/Constants.cs]: This doesn't really show up in the logs. This is a scenario when the EdgeAgent doesn't have its previous records. 
Also, I feel "unknown" is probably just as good as "unknown-module-docker-image" in conveying the info. Let me know if you disagree...

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs (hunk 2 above)
Comment (varunpuranik) [reply edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs]: Sure..

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs [hunk 3]
```diff
@@ -66,7 +64,39 @@ public bool Equals(DockerConfig other)
...
+            CreateContainerParameters createContainerParameters = null;
+            if (!string.IsNullOrWhiteSpace(createOptions) && !createOptions.Equals("null", StringComparison.OrdinalIgnoreCase))
+            {
+                createContainerParameters = JsonConvert.DeserializeObject<CreateContainerParameters>(createOptions);
+            }
+
+            return createContainerParameters ?? new CreateContainerParameters();
+        }
+
+        internal static string ValidateAndGetImage(string image)
```
Comment (myagley) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs]: Is this fixing the issue with appending `latest` tag if no tag is present?

Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs]: Yes, that is correct.

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs [hunk 4]
```diff
@@ -124,5 +138,51 @@ public void TestSerialization()
...
+        [MemberData(nameof(GetTestGetCreateOptionsData))]
+        public void TestGetCreateOptions(string createOptions, CreateContainerParameters expectedCreateOptions)
+        {
+            CreateContainerParameters createContainerParameters = DockerConfig.GetCreateOptions(createOptions);
+
+            Assert.NotNull(createContainerParameters);
+            Assert.True(DockerConfig.CompareCreateOptions(expectedCreateOptions, createContainerParameters));
+        }
+
+        static IEnumerable<object[]> GetTestGetCreateOptionsData()
```
Comment (philipktlin) [inline edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs]: please make this as public; I have PR to upgrade XUnit NuGet package to v2.4.1, which enforces all data provider method should be public.

**Pull request merged**